### PR TITLE
[[ Bug 16568 ]] Make sure native layers are clipped to the bounds of containing groups

### PR DESCRIPTION
**Pull request by livecodeian**
Adds a few MCObject state change messages: geometrychanged, viewportgeometrychanged, OnOpen, OnClose. These are used to manage the native views, which have been moved from MCWidget to MCObject.

### CODE REVIEW

**Comment by livecodefraser** [comment]
Looks like there's a bit of Linux build unhappiness. From my reading of the code, this won't fix the bug on Linux anyway?

**Comment by livecodeian** [comment]
Build & Test bugs now fixed. The Linux & Windows native layers use a different method to clip the layer using a single container rather than nested containers

File: engine/src/object.h [hunk 1]
```diff
@@ -374,7 +380,8 @@ class MCObject : public MCDLlist
 	
 	virtual void timer(MCNameRef mptr, MCParameter *params);
 	virtual uint2 gettransient() const;
-	virtual void setrect(const MCRectangle &nrect);
+	virtual void doSetRect(const MCRectangle &nrect);
+	virtual void setrect(const MCRectangle &p_rect);
```
Comment (runrevmark) [inline engine/src/object.h]: Does this still have to be virtual if it is now a wrapper which prompts a geomchanged?

File: engine/src/object.h [hunk 2]
```diff
@@ -374,7 +380,8 @@ class MCObject : public MCDLlist
 	
 	virtual void timer(MCNameRef mptr, MCParameter *params);
 	virtual uint2 gettransient() const;
-	virtual void setrect(const MCRectangle &nrect);
+	virtual void doSetRect(const MCRectangle &nrect);
```
Comment (runrevmark) [inline engine/src/object.h]: I'd suggest making this 'applyrect' rather than doSetRect, it is a little more consistent than the camel-case version.

**Comment by runrevmark** [comment]
@livecode-vulcan review ok 14e8e88

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 14e8e8882b73bb505acce92308f3242053d2171e

**Comment by livecode-vulcan** [comment]
😎 test success 14e8e8882b73bb505acce92308f3242053d2171e
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1820)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1833)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1814)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1826)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1836)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1839)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1845)

**Pull request merged**